mtt/src/mtt_types.rs

```diff
@@ -0,0 +1,571 @@
+        // Bits 63:36 must be zero
+        const MTT_L2_MBZ_MASK: u64 = 0xffff_fff0_0000_0000;
+        // Bits 35:34
+        const MTT_L2_TYPE_MASK: u64 = 0x000c_0000_0000;
+        // Bits 33:0
+        const MTT_L2_INFO_MASK: u64 = 0x0003_ffff_ffff;
+
+        if (value & MTT_L2_MBZ_MASK) != 0 {
+            return None;
+        }
+        let mtt_l2_type = (value & MTT_L2_TYPE_MASK) >> 34;
+        match mtt_l2_type {
+            // INFO must be 0 for 1GB entries
+            0 if (value & MTT_L2_INFO_MASK) == 0 => Some(MttNonConfidential1G),
+            1 if (value & MTT_L2_INFO_MASK) == 0 => Some(MttConfidential1G),
+            2 => Some(MttL1Dir(value & MTT_L2_INFO_MASK)),
+            3 => Some(Mtt2MBPages(value & MTT_L2_INFO_MASK)),
+            _ => None,
+        }
+    }
+
+    fn raw(&self) -> u64 {
+        match *self {
+            MttNonConfidential1G => 0,
+            MttConfidential1G => 1u64 << 34,
+            MttL1Dir(v) => (2u64 << 34) | v,
+            Mtt2MBPages(v) => (3u64 << 34) | v,
+        }
+    }
+
+    fn confidential_64_mb_range() -> MttL2EntryType {
+        Mtt2MBPages(0xffff_ffff)
+    }
+
+    fn non_confidential_64_mb_range() -> MttL2EntryType {
+        Mtt2MBPages(0)
+    }
+
+    fn mtt_l1_dir(mtt_l1_page: SupervisorPageAddr) -> MttL2EntryType {
+        MttL1Dir(mtt_l1_page.pfn().bits())
+    }
+}
+
+use MttL1EntryType::*;
+impl MttL1EntryType {
+    fn from_raw(value: u64) -> Option<MttL1EntryType> {
+        match value {
+            0 => Some(MttNonConfidential4K),
+            1 => Some(MttConfidential4K),
+            _ => None,
+        }
+    }
+}
+
+/// MTT related errors.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// The MTT L2 entry corresponding to the physical address was invalid.
+    /// This should be treated as an operational error.
+    InvalidL2Entry,
+    /// The MTT L1 entry corresponding to the physical address was invalid (encoding of 11b).
+    /// This should be treated as an operational error.
+    InvalidL1Entry,
+    /// The MTT L2 entries for the specified 1GB range failed to pass the invariant that all
+    /// 16 consecutive entries (each mapping 64MB) must have the same type.
+    ///  This should be treated as an operational error.
+    Invalid1GBRange,
+    /// The caller specified an operation on a 1GB range, but the L2 entry for the physical
+    /// address does not map a 1GB range.
+    Non1GBL2Entry,
+    /// The caller specified a operation on a 64MB range, but the L2 entry for the physical
+    /// address doesn't map a 64MB range.
+    Non64MBL2Entry,
+    /// The caller specified a operation on a 4K page, but the L2 entry for the physical
+    /// address doesn't map a 4K page.
+    Non4KL2Entry,
+    /// The caller specified an invalid physical address (too large for example)
+    InvalidAddress,
+}
+
+/// Holds the result of a MTT operation.
+pub type Result<T> = core::result::Result<T, Error>;
+
```

Comment:
Would a struct like this help better isolate the unsafety? Can wrap it in a ONCE so it's a singleton.

```
/// Struct that holds the base pointer for the MTT.
struct MttBase {
    base: &'static mut [u64: MTT_L2_ENTRY_COUNT],
}

impl MttBase {
    /// Creates a new MTTBase object.
    /// # Safety
    /// The caller must ensure that the `base_addr` is a valid pointer to a 8MB MTT region, that is not aliased elsewhere.
    pub unsafe fn new(base_addr:u64) -> Self {
        let base = core::slice::from_raw_parts_mut(base_addr as *mut u64, MTT_L2_ENTRY_COUNT);
        Self { base }
    }
}

```

Probably should use a VolatileSlice instead of a slice though to guarantee nothing gets elided because the results are never used.

Reply:
Possibly, but note that the MTT is allocated by the trusted-platform FW, so there's no actual static allocation for it in Salus. So, it's similar to the FDT and other structures that are passed by FW (except that in this case, we'll likely need to just read it from a platform specific CSR and cross-check that it's marked as reserved in the DT).

Reply:
the 'static lifetime above is saying that it's memory that lives for the life of salus. Is that true? Does the firmware allocate the table and then hand that memory to salus for as long as the system is booted?

Reply:
That is correct. The proposed mechanism is for the platform FW to allocate the memory upfront, and mark it as reserved in the DT (with some custom attribute to help identify it as such). An alternative is for Salus to carve out the memory as part of its bootstrap, and then reserve it in its own internal map before passing on the DT to the host. In either case, the memory must remain allocated until the platform reboots.

Reply:
My goal is to move the unsafe code from regular execution paths to one-time setup. Do you think that is possible? If we establish that salus has unique ownership of the mtt memory once, then we can safely access it for the lifetime of salus.

Reply:
I think we can get pretty close. The wildcard is that unlike the fixed size MTT L2 entry, L1 pages aren't statically sized. However, given that they must be in contiguous memory, we likely can establish the size once using the DT, and then return pages from the L1 pool at runtime. That way, each L1 page is just scoped as a `&'static mut [u64; 512]` or similar. I have made the changes, but am still trying it out in the playground because I need to convince myself that I got the slice de-referencing correct.

Reply:
Sorry, took me a while, but believe that I have an abstraction that we can be happy with. Basically, it will use unsafe {} once during init, and everything else will be safe. Let me verify that it works as expected and post an update tomorrow.

Reply:
Done....PTAL.